js/tooltips.js

```diff
@@ -0,0 +1,153 @@
+    this.tooltipsView.$el.css({
+      'left': `${$(window).scrollLeft()}px`,
+      'top': `${$(window).scrollTop()}px`
+    });
+
+    // render the tooltip
+    this.show(id);
+
+    // calculate optimum position
+    const availableWidth = $('html')[0].clientWidth;
+    const availableHeight = $('html')[0].clientHeight;
+    let tooltipsWidth = this.tooltipsView.$el.width();
+    let tooltipsHeight = this.tooltipsView.$el.height();
+    const scrollTop = $(window).scrollTop();
+    const scrollLeft = $(window).scrollLeft();
+    const canAlignTop = targetBoundingRect.top - tooltipsHeight >= 0;
+    const canAlignBottom = targetBoundingRect.bottom + tooltipsHeight + scrollTop < availableHeight;
+    const canAlignLeft = targetBoundingRect.left - tooltipsWidth >= 0;
+    const canAlignRight = targetBoundingRect.right + tooltipsWidth + scrollLeft < availableWidth;
+    const canAlignBottomRight = canAlignBottom && canAlignRight;
+    const canAlignTopRight = canAlignTop && canAlignRight;
+    const canAlignBottomLeft = canAlignBottom && canAlignLeft;
+    const canAlignTopLeft = canAlignTop && canAlignLeft;
+
+    const alignBottomRight = () => {
+      console.log('alignBottomRight');
+      this.tooltipsView.$el.css({
+        'left': `${targetBoundingRect.right + scrollLeft}px`,
+        'top': `${targetBoundingRect.bottom + scrollTop}px`
+      });
+    }
+
+    const alignTopRight = () => {
+      console.log('alignTopRight');
+      this.tooltipsView.$el.css({
+        'left': `${targetBoundingRect.right + scrollLeft}px`,
+        'top': `${targetBoundingRect.top - tooltipsHeight + scrollTop}px`
+      });
+    }
+
+    const alignBottomLeft = () => {
+      console.log('alignBottomLeft');
+      this.tooltipsView.$el.css({
+        'left': `${targetBoundingRect.left - tooltipsWidth + scrollLeft}px`,
+        'top': `${targetBoundingRect.bottom + scrollTop}px`
+      });  
+    }
+
+    const alignTopLeft = () => {
+      console.log('alignTopLeft');
+      this.tooltipsView.$el.css({
+        'left': `${targetBoundingRect.left - tooltipsWidth + scrollLeft}px`,
+        'top': `${targetBoundingRect.top - tooltipsHeight + scrollTop}px`
+      });
+    }
+
+    if (canAlignBottomRight) {
+      alignBottomRight();
+    } else if (canAlignTopRight) {
+      alignTopRight();
+    } else if (canAlignBottomLeft) {
+      alignBottomLeft();
+    } else if (canAlignTopLeft) {
+      alignTopLeft();
+    } else {
+      // find the 'corner' with the most space
+      const isTopPreferred = availableHeight - (targetBoundingRect.bottom + tooltipsHeight) < targetBoundingRect.top - tooltipsHeight;
+      const isLeftPreferred = availableWidth - (targetBoundingRect.right + tooltipsWidth) < targetBoundingRect.left - tooltipsWidth;
+
+      if (isTopPreferred) {
+        isLeftPreferred ? alignTopLeft() : alignTopRight();
+      } else {
+        isLeftPreferred ? alignBottomLeft() : alignBottomRight();
+      }
+    }
```

Comment:
```suggestion
    if (canAlignBottomRight) return alignBottomRight();
    if (canAlignTopRight) return alignTopRight();
    if (canAlignBottomLeft) return alignBottomLeft();
    if (canAlignTopLeft) return alignTopLeft();

    // find the 'corner' with the most space
    const isTopPreferred = availableHeight - (targetBoundingRect.bottom + tooltipsHeight) < targetBoundingRect.top - tooltipsHeight;
    const isLeftPreferred = availableWidth - (targetBoundingRect.right + tooltipsWidth) < targetBoundingRect.left - tooltipsWidth;

    if (isTopPreferred && isLeftPrefered) return alignTopLeft();
    if (isTopPrefered) return alignTopRight();
    if (isLeftPrefered) return alignBottomLeft();
    alignBottomRight();
```

Move all positioning stuff to ToolTipView